File: 2020/12/27/1182.py
n, s = map(int, input().split())
number = list(map(int, input().split()))
visited = [False] *n
stack = [] 
ans = 0
def dfs(num):
    if(len(stack) != 0):
        sum = 0
        for iter in stack:
            sum += iter
        if(sum == s):
            global ans
            ans += 1

    if(len(stack) !=n ):
        for i in range(num+1, n):
            if visited[i] is False:
                visited[i] = True
                stack.append(number[i])
                dfs(i)
                stack.pop()
                visited[i] = False
    else:
        return

dfs(-1)
print(ans)

File: 2020/12/27/1182.cpp
// #include <cstdio>
// #include <vector>
// using namespace std;
// int n, s;
// int number[21] ={};
// bool visited[21] ={false};
// int answer =0;
// vector<int> stack;
// void dfs(int num){
//     if(stack.size() != 0){
//         int sum = 0;
//         for(int i = 0; i< stack.size();i++){
//             sum += stack[i];
//         }
//         if(sum == s){
//             answer++;
//         }
//     }
//     if(stack.size() != n){
//         for(int i = num+1 ; i< n;i++){
//             if(!visited[i]){
//                 visited[i] =true;
//                 stack.push_back(number[i]);
//                 dfs(i);
//                 visited[i] =false;
//                 stack.pop_back();
//             }
//         }
//     }
// }

// int main(){
//     scanf("%d %d",&n, &s);
//     for(int i = 0; i < n; i++){
//         scanf("%d", &number[i]);
//     }
//     dfs(-1);
//     printf("%d\n",answer);
//     return 0;
// }
#include <cstdio>
using namespace std;

int n, s, sum, cnt;
int arr[20];

void dfs(int i, int sum){
    if (i== n) return;
    if (sum+arr[i] == s) cnt++;

    dfs(i+1, sum);
    dfs(i+1, sum +arr[i]);
}

int main(){
    scanf("%d %d", &n, &s);
    for(int i =0; i< n; i++) scanf("%d", &arr[i]);
    cnt = 0;
    dfs(0, 0);

    printf("%d\n", cnt);
    return 0;
}